### PR TITLE
Remove the 'set_language' parameter when 'always_set_cookie' is enabled (master)

### LINKED ISSUE #362: 'set_language' request attribute is not needed when the Language Tool is configured to set the cookie always
The 'set_language' parameter that the language selector adds in the request is only needed to force the language negotiation machinery to set the language cookie to the user. See https://github.com/plone/plone.i18n/blob/master/plone/i18n/negotiate/negotiate.py#L34

This cookie is a session cookie: https://github.com/plone/plone.i18n/blob/bd6c6fdf274acd23fe46010c173c111c3e501e53/plone/i18n/utility.py#L225

So the cookie is only needed in this scenario:

- A user goes to a content in english
- He changes the language to spanish (follows the URL with set_language=es)
- He goes to the root of the site (www.domain.com) -> Plone negotiates the languages according to the cookie and redirects the user accordingly to www.domain.com/es/

Nowadays the language negotiator gives more preference in the negotiation procedure to the content of the language itself.

There is a setting in the language control panel "set the cookie always" which sets the cookie always, regardless the presence of "set_language".

So we can tune the language change viewlet to remove that set_language parameter from the request if that "set the cookie always" settings is on, avoiding creating the set_language=XX urls,  and avoiding extra robot traffic with those "different" urls with the request parameter. If the user lands in a page and has no cookie the "set the cookie always" will force the machinery to set him the cookie so he will keep browsing in the correct language.

@libargutxi @garaolaza

### PR DESCRIPTION
**Pull request by erral**
Fixes #362 

@libargutxi @garaolaza

### CODE REVIEW

**Comment by erral** [comment]
@jenkins-plone-org please run jobs

File: src/plone/app/multilingual/tests/test_selector.py [hunk 1]
```diff
@@ -930,3 +931,146 @@ def assertUrlsEqual(self, url1, url2):
...
+                    },
+                    'selected': False,
+                    'translated': True,
+                },
+            ],
+        )
+
+
+    def test_set_language_is_not_present_when_always_set_cookie_is_set(self):
+        """ test the presence of set_language parameter in the urls created in the language selector"""
```
Comment (ale-rt) [inline src/plone/app/multilingual/tests/test_selector.py]: Should be `...test the absence of set_language...`

File: src/plone/app/multilingual/tests/test_selector.py [hunk 2]
```diff
@@ -930,3 +931,146 @@ def assertUrlsEqual(self, url1, url2):
...
+        self.settings.set_cookie_always = False
+        self.selector_viewlet = LanguageSelectorViewlet(
+            self.portal['en'], self.request, None, None
+        )
+        self.selector_viewlet.update()
+
+        selector_languages = self.selector_viewlet.languages()
+        tg = ITG(self.portal['en'])
+
+        self.assertEqual(
```
Comment (ale-rt) [inline src/plone/app/multilingual/tests/test_selector.py]: I suggest using `self.assertListEqual` (not a blocker)

File: src/plone/app/multilingual/tests/test_selector.py [hunk 3]
```diff
@@ -930,3 +931,146 @@ def assertUrlsEqual(self, url1, url2):
         self.assertEqual(parsed_url_1[2], parsed_url_2[2])
         self.assertEqual(parsed_url_1[3], parsed_url_2[3])
         self.assertEqual(parse_qs_1, parse_qs_2)
+
+
+class TestLanguageSelectorSetLanguage(unittest.TestCase):
+    layer = PAM_FUNCTIONAL_TESTING
```
Comment (ale-rt) [inline src/plone/app/multilingual/tests/test_selector.py]: I think an Integration test is enough (and much faster)

File: src/plone/app/multilingual/tests/test_selector.py [hunk 4]
```diff
@@ -930,3 +931,146 @@ def assertUrlsEqual(self, url1, url2):
...
+
+    def setUp(self):
+        # Set test variables
+        self.portal = self.layer['portal']
+        self.portal_url = self.portal.absolute_url()
+        self.request = self.layer['request']
+        alsoProvides(self.layer['request'], IPloneAppMultilingualInstalled)
+
+        # Setup testbrowser
+        self.browser = Browser(self.layer['app'])
```
Comment (ale-rt) [inline src/plone/app/multilingual/tests/test_selector.py]: self.browser is not needed, right?

File: src/plone/app/multilingual/tests/test_selector.py [hunk 5]
```diff
@@ -930,3 +931,146 @@ def assertUrlsEqual(self, url1, url2):
...
+
+class TestLanguageSelectorSetLanguage(unittest.TestCase):
+    layer = PAM_FUNCTIONAL_TESTING
+
+    def setUp(self):
+        # Set test variables
+        self.portal = self.layer['portal']
+        self.portal_url = self.portal.absolute_url()
+        self.request = self.layer['request']
+        alsoProvides(self.layer['request'], IPloneAppMultilingualInstalled)
```
Comment (ale-rt) [inline src/plone/app/multilingual/tests/test_selector.py]: Isn't the fixture already taking care of assigning `IPloneAppMultilingualInstalled`? I think this is not needed, can you verify that?

File: src/plone/app/multilingual/tests/test_selector.py [hunk 6]
```diff
@@ -930,3 +931,146 @@ def assertUrlsEqual(self, url1, url2):
...
+        self.settings.set_cookie_always = True
+        self.selector_viewlet = LanguageSelectorViewlet(
+            self.portal['en'], self.request, None, None
+        )
+        self.selector_viewlet.update()
+
+        selector_languages = self.selector_viewlet.languages()
+        tg = ITG(self.portal['en'])
+
+        self.assertEqual(
```
Comment (ale-rt) [inline src/plone/app/multilingual/tests/test_selector.py]: I suggest using self.assertListEqual (not a blocker)

File: src/plone/app/multilingual/tests/test_selector.py (hunk 1 above)
Comment (erral) [reply src/plone/app/multilingual/tests/test_selector.py]: indeed... I copied over the method and did not change the comment :face_with_head_bandage:

File: src/plone/app/multilingual/tests/test_selector.py (hunk 5 above)
Comment (erral) [reply src/plone/app/multilingual/tests/test_selector.py]: It should be. But if you check it, other testcases also do the alsoProvides dance.

Comment (erral) [reply src/plone/app/multilingual/tests/test_selector.py]: I verified and it is not needed.

We can do it for all tests in pam (in a later PR)

**Comment by erral** [comment]
Thank your your review @ale-rt 

@jenkins-plone-org

**Comment by erral** [comment]
@jenkins-plone-org please run tests

**Comment by erral** [comment]
@jenkins-plone-org please run jobs

**Comment by ale-rt** [comment]
I restarted the failing job, now waiting for a node to be ready

**Pull request merged**